applications/FluidDynamicsApplication/custom_utilities/fluid_auxiliary_utilities.cpp

```diff
@@ -343,6 +373,113 @@ FluidAuxiliaryUtilities::ModifiedShapeFunctionsFactoryType FluidAuxiliaryUtiliti
     }
 }
 
+void FluidAuxiliaryUtilities::InternalCreateSlipMultiPointConstraints(
+    ModelPart& rModelPart,
+    const Variable<array_1d<double,3>>& rUnknownVar,
+    const Flags& rSlipFlag,
+    const std::function<double(const NodeType&, const array_1d<double,3>&)>& rWallNormalVelocityCalculator)
+{
+// Get MPI communicators
+    const auto& r_comm = rModelPart.GetCommunicator();
+    const auto& r_data_comm = r_comm.GetDataCommunicator();
+    KRATOS_ERROR_IF(r_comm.IsDistributed()) << "Slip boundary condition through MPCs is not fully MPI-compatible yet." << std::endl;
+
+    // Remove all the MPCs tagged with the slip flag
+    rModelPart.RemoveMasterSlaveConstraints(rSlipFlag);
+
+    // Get the maximum MPCs id
+    int max_id(0);
+    if (r_comm.LocalMesh().NumberOfMasterSlaveConstraints() != 0) {
+        max_id = block_for_each<MaxReduction<IndexType>>(r_comm.LocalMesh().MasterSlaveConstraints(), [&](MasterSlaveConstraint& rMPC){
+            return rMPC.Id();
+        });
+    }
+    max_id = r_data_comm.MaxAll(max_id);
+
+    // Create the slip nodal MPCs
+    //TODO: We should make the id increment MPI compatible
+    if (r_comm.LocalMesh().NumberOfNodes() != 0) {
+        const IndexType dim = rModelPart.GetProcessInfo()[DOMAIN_SIZE];
+
+        // Create a map with the component unknowns
+        // Note that this is done in order to access the KratosComponents once
+        std::map<IndexType, const Variable<double>*> unknowns_map;
+        unknowns_map[0] = &KratosComponents<Variable<double>>::Get(rUnknownVar.Name()+"_X");
+        unknowns_map[1] = &KratosComponents<Variable<double>>::Get(rUnknownVar.Name()+"_Y");
+        if (dim == 3) {
+            unknowns_map[2] = &KratosComponents<Variable<double>>::Get(rUnknownVar.Name()+"_Z");
+        }
+
+        // Allocate auxiliary containers
+        array_1d<double,3> unit_normal;
+        DofPointerVectorType slave_dofs;
+        DofPointerVectorType master_dofs;
+        VectorType constant_vector(1);
+        MatrixType relation_matrix(1,dim-1);
+        ModelPart::MasterSlaveConstraintContainerType mpcs_container;
+        slave_dofs.reserve(1);
+        master_dofs.reserve(dim-1);
+        mpcs_container.reserve(rModelPart.NumberOfNodes());
+
+        // Loop nodes to create the slip MPCs
+        for (auto& r_node : rModelPart.Nodes()) {
+            if (r_node.Is(rSlipFlag)) {
+                // Calculate the nodal unit normal
+                unit_normal = r_node.FastGetSolutionStepValue(NORMAL);
```

Comment:
here you are assuming that NORMAL is in the historical database. you sure about this?